### ISSUE TITLE
[gcc] Missing/wrong binary names for ar, nm, ranlib, strip

### ISSUE THREAD

**Issue by totszwai** [msys2/MINGW-packages#19791]
### Description / Steps to reproduce the issue

When installing the group [mingw-w64-ucrt-x86_64-toolchain](https://packages.msys2.org/groups/mingw-w64-ucrt-x86_64-toolchain), the following binutils are missing (expecting the exact naming below):
- x86_64-w64-mingw32-ar.exe
- x86_64-w64-mingw32-nm.exe
- x86_64-w64-mingw32-ranlib.exe
- x86_64-w64-mingw32-strip.exe

Why is it that the UCRT64 variants contains "gcc" in the name (and strip is missing)?
```
/ucrt64/bin/x86_64-w64-mingw32-gcc-ar.exe
/ucrt64/bin/x86_64-w64-mingw32-gcc-nm.exe
/ucrt64/bin/x86_64-w64-mingw32-gcc-ranlib.exe
```

I was going to perhaps try using the cross-binutils, but then the naming is all messed up too. Why is [mingw-w64-cross-binutils](https://packages.msys2.org/package/mingw-w64-cross-binutils?repo=msys&variant=x86_64) installing them in the `/opt` and have a completely different naming convention?:

![image](https://github.com/msys2/MINGW-packages/assets/4440365/54a60cd0-4c94-4180-826a-81e89d4c3197)

There seem to be some major naming convention issue here between mingw64 and ucrt64... we have the following groups but they are all doing things differently:
- [mingw-w64-cross-toolchain](https://packages.msys2.org/groups/mingw-w64-cross-toolchain)
  - Contains all flavors, BUT it follows a weird installation by putting them under `/opt` with different naming scheme?
- [mingw-w64-x86_64-toolchain](https://packages.msys2.org/groups/mingw-w64-x86_64-toolchain)
  - Contains only MINGW64, uses prefix `x86_64-w64-mingw32-`, putting them under `/mingw64`
- [mingw-w64-ucrt-x86_64-toolchain](https://packages.msys2.org/groups/mingw-w64-ucrt-x86_64-toolchain)
  - Contains only UCRT64, ALSO uses prefix `x86_64-w64-mingw32-`, putting them under `/ucrt64`, but missing several binaries such as ar, nm, ranlib, etc.

Which packages/group are we supposed to be installing when doing cross compiling?
mingw-w64-cross-toolchain? mingw-w64-ucrt-x86_64-toolchain? mingw-w64-x86_64-toolchain?

### Expected behavior

Expecting no matter which group of toolchains I install, they all follows the same deployment/naming convention... either:

- Putting them ALL under `/<environment>`. ALL shares the SAME prefixes.
OR
- Putting them ALL under `/opt`, following the SAME naming convention prefixes such as `x86_64-w64-mingw32<environment>-`

Where `environment` would be mingw64, clang64, ucrt64 etc

### Actual behavior

Currently, there are some weird behaviors and missing executables going on depending which group of packages you install.

Using mingw-w64-ucrt-x86_64-toolchain? Missing ar, nm, ranlib, strip
Using mingw-w64-cross-toolchain? They all go to `/opt` but with a unique/weird naming convention on the binaries

### Verification

- [X] I have verified that my MSYS2 is up-to-date before submitting the report (see https://www.msys2.org/docs/updating/)

### Windows Version

MINGW64_NT-10.0-22621

### MINGW environments affected

- [X] MINGW64
- [X] MINGW32
- [X] UCRT64
- [ ] CLANG64
- [ ] CLANG32
- [ ] CLANGARM64

**Comment by Biswa96** [comment]
> When installing the group [mingw-w64-ucrt-x86_64-toolchain](https://packages.msys2.org/groups/mingw-w64-ucrt-x86_64-toolchain), the following binutils are missing (expecting the exact naming below)

Those prefixed programs are not required for native gcc toolchain for targeting native Windows platform.

> I was going to perhaps try using the cross-binutils, but then the naming is all messed up too.

The cross binutils is used to cross compile programs from msys2 environment to mingw environment. It is not required for targeting native Windows platform.

tl;dr Just follow these steps https://www.msys2.org/

**Comment by totszwai** [comment]
@Biswa96 I am trying to cross compiling for Windows application, so which group should I be using? I am trying to cross compile it under Ubuntu 22.04 to build it whether is mingw64 or ucrt64.

**Comment by Biswa96** [comment]
> I am trying to cross compile it under Ubuntu 22.04 to build it whether is mingw64 or ucrt64.

You have to use the cross compiler toolchain provided by Ubuntu. In  your case host = Linux and target = Windows.

**Comment by totszwai** [comment]
@Biswa96 When using the MSYS2 UCRT64 shell, which packages should I be using/updating or do I just use it "as-is"? Because in our Makefile, I have some build tool mappings, and we map those `c++`, `gcc`, `g++` against the prefixes `x86_64-w64-mingw32`. So you are saying for stuff like `ar`, `nm`, `ranlib` etc, I should not map those?

**Comment by Biswa96** [comment]
> So you are saying for stuff like `ar`, `nm`, `ranlib` etc, I should not map those?

Yes, you should not map those for native compilation.

**Comment by totszwai** [comment]
> > So you are saying for stuff like `ar`, `nm`, `ranlib` etc, I should not map those?
> 
> Yes, you should not map those for native compilation.

@Biswa96 Wait... no that's not exactly true. So I am building some 3rd party libraries such as zlib, openssl or any other lib, when I am passing the argument `--cross-compile-prefix=` it would look for `x86_64-w64-mingw32-ar`. These aren't my Makefile, they are from the 3rd party libraries, and so I believe ming64/ucrt64/etc must supply the correct binaries with the proper naming convention.

We have our own patches applied to the zlib/openssl etc, but we are not touching the original Makefile.

**Comment by Biswa96** [comment]
The [mingw-w64-ucrt-x86_64-toolchain](https://packages.msys2.org/groups/mingw-w64-ucrt-x86_64-toolchain) is not a cross compiler toolchain. So, `--cross-compile-prefix=` option is not required.

**Comment by totszwai** [comment]
> The [mingw-w64-ucrt-x86_64-toolchain](https://packages.msys2.org/groups/mingw-w64-ucrt-x86_64-toolchain) is not a cross compiler toolchain. So, `--cross-compile-prefix=` option is not required.

Sooooooo if I understood you correctly, when using MINGW64 or UCRT64, I should just be using all the regular `c++`, `g++`, `gcc` and ignore the other ones with prefixes? 

![image](https://github.com/msys2/MINGW-packages/assets/4440365/0000e825-1bda-4da8-9704-6b0eb12de817)

How would you determine which environment you are in? Whether is a cross-compile mingw64 or MSYS2 MINGW64 native shell...? Because now it sounds like there are 2 different ways to compile the same thing.

- IF, cross compiling MINGW (ex: under Ubuntu), then I need to append the prefix
- IF, using MSYS2 native shell but STILL MINGW, then no prefix...

In the 2 cases above, BOTH compiling with mingw for Windows but with 2 different ways.... 

Sorry for asking so many questions, I am confused and currently trying to transition over to use UCRT64.

**Comment by totszwai** [comment]
I believe this should be reopen or at least take a deeper look into it?

**Comment by mmuetzel** [comment]
Usually `Makefile`s use variables like `CC`, `AR`, ... with some reasonable (non-prefixed) default values.
If users would like to cross-compile, they would need to set those variables. E.g., if they'd like to cross-build from Linux, they'd build with `make CC=x86_64-w64-mingw32-gcc AR=x86_64-w64-mingw32-ar`.
Users building on Linux for Linux would not set those variables. Same for users building on Windows for Windows...
